### ISSUE TITLE
terraform_module_pinned_source rule emits unclear error for un-parseable URL

### ISSUE THREAD

**Issue by mjneuharth** [terraform-linters/tflint#1196]
We have an project in AzureDevops which was created with a space in the name, "My Project". When referencing repos within that project, tflint fails on the pinned_source rule, seemingly misreading the end of line.

For this bit of terraform,

```hcl
module "rg" {
  source   = "git::https://myorg:#{ terraformAzureDevOpsPAT }#@dev.azure.com/myorg/My%20Project/_git/terraform-az-rg?ref=0.1.0"
  name     = "rg-${local.application}-${var.environment}-${var.fpbuild_id}"
  location = var.location
  tags     = local.default_tags
}
```

We see this error message,

```
2021-08-24 13:01:40 [ERROR]   Found errors in [tflint] linter!
2021-08-24 13:01:40 [ERROR]   Error code: 3. Command output:
------
1 issue(s) found:

Warning: Module source "git::https://myorg:#{ terraformAzureDevOpsPAT }#@dev.azure.com/myorg/My%!P(MISSING)roject/_git/terraform-az-rg?ref=0.1.0" is not pinned (terraform_module_pinned_source)

  on /tmp/lint/main.tf line 2:
   2:   source   = "git::https://myorg:#{ terraformAzureDevOpsPAT }#@dev.azure.com/myorg/My%20Plus/_git/terraform-az-rg?ref=0.1.0"

Reference: https://github.com/terraform-linters/tflint/blob/v0.31.0/docs/rules/terraform_module_pinned_source.md
```

## Version

```console
> tflint -v
TFLint version 0.31.0
> terraform -v
Terraform v0.14.10
```

**Comment by bendrucker** [comment]
> We have an project in AzureDevops which was created with a space in the name, "My Project".

This is unrelated, as you'll see if you just rename `My%20Project` to `my-project` and re-run. 

> seemingly misreading the end of line.

Here's a bit of insight into how your input is parsed:

https://play.golang.org/p/gxSmHAvkvt7

```
&url.URL{
    Scheme:      "https",
    Opaque:      "",
    User:        (*url.Userinfo)(nil),
    Host:        "myorg:",
    Path:        "",
    RawPath:     "",
    ForceQuery:  false,
    RawQuery:    "",
    Fragment:    "{ terraformAzureDevOpsPAT }#@dev.azure.com/myorg/My Project/_git/terraform-az-rg?ref=0.1.0",
    RawFragment: "{ terraformAzureDevOpsPAT }#@dev.azure.com/myorg/My%20Project/_git/terraform-az-rg?ref=0.1.0",
}
```

https://en.wikipedia.org/wiki/URI_fragment

> In URIs, a hash mark # introduces the optional fragment near the end of the URL. The generic RFC 3986 syntax for URIs also allows an optional query part introduced by a question mark ?. In URIs with a query and a fragment, the fragment follows the query. 

I'm adding a bit of validation to emit a clearer issue when the rule cannot understand the provided URL but knows it's a `git` or `hg` URL. 

Presumably you have some sort of custom step replacing either the literal `#{ terraformAzureDevOpsPAT }#` or treating `#{}` as a template construction. You could either invoke this with a fake token prior to running TFLint or you could use an alternative interpolation format that is URL-safe.

**Comment by mjneuharth** [comment]
Terrific response, @bendrucker. Thanks so much for all of this. 

This was local testing and when I get it into the pipeline that substitution step will get done before tflint runs. 

Thanks!
